extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/interceptor/TransactionalInterceptorBase.java

```diff
@@ -124,8 +124,7 @@ protected Object invokeInOurTx(InvocationContext ic, TransactionManager tm, Runn
             // handle asynchronously if not throwing
             if (!throwing && ret != null) {
                 ReactiveTypeConverter<Object> converter = null;
-                if (ret instanceof CompletionStage == false
-                        && ret instanceof Publisher == false) {
+                if (ret instanceof CompletionStage == false && !ret.getClass().equals(Publisher.class)) {
```

Comment:
Shouldn't it be:

```java
if(ret instanceof CompletionStage == false
 && (ret instanceof Publisher == false || !ic.getMethod().getReturnType() == Publisher.class))
```

You need to check the method's declared return type, not the instance exact type, which can never be `Publisher` since it's an interface.
Note that I wonder if we don't have the same issue with `CompletionStage` and `CompletableFuture`. Could you add a test to see what happens if an intercepted method is declared to return a `CompletableFuture`?

Reply:
Adding the test right now.

Reply:
I fixed the condition.